File: demos/javascript/dsa/sorting/quick_sort.js
function quickSort(arr) {
  if (arr.length <= 1) return arr;
  const pivot = arr[Math.floor(arr.length / 2)];
  const left = arr.filter(x => x < pivot);
  const middle = arr.filter(x => x === pivot);
  const right = arr.filter(x => x > pivot);
  return [...quickSort(left), ...middle, ...quickSort(right)];
}

console.log("=== JavaScript QuickSort & V8 TimSort ===");
const data = [64, 25, 12, 22, 11];
const sorted = quickSort(data);
console.assert(JSON.stringify(sorted) === JSON.stringify([11, 12, 22, 25, 64]), "Sort failed");
console.log("Sorted:", sorted);
console.log("JavaScript QuickSort tests passed successfully.");

File: demos/python/dsa/sorting/quick_sort.py
def quick_sort(arr: list[int]) -> list[int]:
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

def main():
    print("=== Python Functional QuickSort & TimSort ===")
    data = [64, 25, 12, 22, 11]
    sorted_data = quick_sort(data)
    assert sorted_data == [11, 12, 22, 25, 64]
    assert sorted(data) == sorted_data

    print(f"Sorted data: {sorted_data}")
    print("Python QuickSort tests passed successfully.")

if __name__ == "__main__":
    main()
